baseband_tasks/pulse_gate.py

```diff
@@ -0,0 +1,162 @@
+    tol : `~astropy.units.Quantity`
+        The tolarence of the pulse phaes gating in the units of cycle. Tolarence
+        can not be bigger than 10% of the gate size. Default is 5% of the given
+        gate size.
+    pulse_period: `~astropy.unit.Quantity` object, optional
+        Input pulse period. If not given, it will search from the input
+        callable `phase` object. If given, it has to be in the unit of time.
+
+    """
+    def __init__(self, ih, phase, gate, tol=None, pulse_period=None, dtype=None):
+        self.ih = ih
+        if pulse_period is None: # Get pulse period from phase class
+            try:
+                self.pulse_period = 1.0 / phase.apparent_spin_freq(ih.start_time)
+            except AttributeError:
+                raise ValueError("Can not find pulse period from the `phase` "
+                                 "class. Please provide a valid"
+                                 " `pulse_period`.")
+        else:
+            self.pulse_period = pulse_period.to(u.s)
+
+        self.dtype = getattr_if_none(ih, 'dtype', dtype)
+        # Use the phase searching code in the Integrate class.
+        self.phase = phase
+        self.gate = self.verify_gate(gate)
+        self.samples_per_period = int(self.pulse_period * self.ih.sample_rate)
+        self.phase_per_sample = 1.0 * u.cycle / self.samples_per_period
+        # Check the tolerence range, tolerance has to smaller than 10% of gate size
+        if tol is None:
+            tol =  (self.gate[1] - self.gate[0]) / 20
+        assert tol > 0
+        assert tol < (self.gate[1] - self.gate[0]) / 10
+        self.tol = tol
+        # Compute the number of samples in the tolerance phase, which means, the
+        # Gated data will be accurate to `self.tol_sample` of samples.
+        self.tol_sample = tol / self.phase_per_sample
+        if self.tol_sample < 1:
+            warnings.warn("Tolarence is smaller than one input time sample. "
+                          "Use one time sample as the tolarence and the edge of"
+                          "the gate will not as accurate as requested.")
+            self.tol_sample = np.ceil(self.tol_sample)
+            self.tol = self.tol_sample * self.phase_per_sample
+        self.tol_sample = self.tol_sample.astype(int)
+        self.pulse_offset = 0
+        self.gate_offsets = self.get_gate_offsets()
+
+    def read(self, gate_index=None):
+        """Read the next pulse.
+
+        Parameter
+        ---------
+        pulse_index: int, optional
+            The index of gate to read. If not provide, it reads the next pulse
+            from the current pulse_offset.
+        """
+        if gate_index is None:
+            gate_index = self.pulse_offset
+
+        if gate_index >= len(self.gate_offsets[0]):
+            raise ValueError("The requested gate index is beyond the total data"
+                             "stream.")
+
+        gsh = GetSlice(self.ih, slice(self.gate_offsets[0][gate_index],
```

Comment:
Why not make a new `get_slice` method that returns this slice (everything from the top of this method to here) and then use that for `read`?